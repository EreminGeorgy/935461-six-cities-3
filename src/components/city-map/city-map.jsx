import React, {useEffect, useRef} from "react";
import leaflet from "leaflet";
import PropTypes from 'prop-types';

const MAP_CONFIG = {
  zoom: 12,
  zoomControl: false,
  marker: true
};

const ICON_CONFIG = {
  iconUrl: `img/pin.svg`,
  iconSize: [30, 40]
};

export const CityMap = (props) => {

  const mapRef = useRef(null);
  const {offers, city} = props;

  const mapConfig = Object.assign({}, MAP_CONFIG, {center: city.location})

  useEffect(() => {
    if (!mapRef.current) {
      return;
    }

    window.map = leaflet.map(mapRef.current, mapConfig);

    window.map.setView(mapConfig.center, mapConfig.zoom);

    leaflet
    .tileLayer(`https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`, {
      detectRetina: true,
      attribution: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
    }).addTo(window.map);

    // return () => {
    //   map = null;
    // };
  });

  useEffect(() => {

    const handleAddPinOnMap = (offerCords) => {
      const icon = leaflet.icon(ICON_CONFIG);

      leaflet
        .marker(offerCords, {icon})
        .addTo(window.map);
    };

    for (let i = 0; i < offers.length; i++) {
      handleAddPinOnMap(offers[i].location[i]);
    }
  }, [offers]);

  return (
    <section className="cities__map">
      <div id="map" ref={mapRef} style={{height: `800px`}}></div>
    </section>
  );
};

CityMap.propTypes = {
  offers: PropTypes.arrayOf(PropTypes.shape({
    location: PropTypes.array.isRequired,
  })).isRequired,
};
