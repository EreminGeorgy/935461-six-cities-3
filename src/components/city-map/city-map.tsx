import * as React from "react";
import {useEffect, useRef} from "react"
import * as leaflet from "leaflet";
import {Offer, City} from "../../types";


const MAP_CONFIG = {
  zoom: 12,
  zoomControl: false,
  marker: true,
};

const ICON_CONFIG = {
  iconSize: [30, 40],
};

const PIN_TYPE = {
  ACTIVE: `img/pin-active.svg`,
  DISABLED: `img/pin.svg`,
};

const NEIGHBOURS = 3;

interface Props {
  offers: Offer[];
  city: City;
  activeCard: number;
  path?: string;
  currentOfferCoords?: number[];
}

export const CityMap: React.FunctionComponent<Props> = (props: Props) => {

  const {offers, city, activeCard, path, currentOfferCoords} = props;
  let mapRef = useRef(null);

  const mapConfig = {
    ...MAP_CONFIG,
    center: city.locations,
    layers: [
      leaflet.tileLayer(`http://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`, {
        detectRetina: true,
        attribution: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
      }),
    ]
  };

  useEffect(() => {
    if (!mapRef.current) {
      mapRef.current = leaflet.map(`mapId`, mapConfig);
    }
    mapRef.current.setView(mapConfig.center, mapConfig.zoom);

  }, [city]);

  useEffect(() => {
    const handleAddPinOnMap = (offerCords, id) => {
      const icon = leaflet.icon({...ICON_CONFIG, iconUrl: `${path ? path : ``}${activeCard === id ? PIN_TYPE.ACTIVE : PIN_TYPE.DISABLED}`});
      leaflet
        .marker(offerCords, {icon})
        .addTo(mapRef.current);
    };

    for (let i = 0; i < offers.length; i++) {
      handleAddPinOnMap(offers[i].locations, offers[i].id);
    }
  }, [offers, activeCard]);

  useEffect(() => {

    if (currentOfferCoords) {
      const iconCurrent = leaflet.icon({...ICON_CONFIG, iconUrl: `${path}${PIN_TYPE.ACTIVE}`});
      leaflet
        .marker(currentOfferCoords, {icon: iconCurrent})
        .addTo(mapRef.current);
    }
  }, [currentOfferCoords]);

  return <section className={offers.length > NEIGHBOURS ? `cities__map` : `property__map`}>
    <div id="mapId" style={{height: `580px`}}></div>
  </section>;
};
