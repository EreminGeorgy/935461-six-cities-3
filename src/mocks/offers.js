const shortid = require(`shortid`);

// const MOCKS_TOTAL = 10;
// const IMAGES = [
//   `img/apartment-01.jpg`,
//   `img/apartment-02.jpg`,
//   `img/apartment-01.jpg`,
//   `img/apartment-03.jpg`,
//   `img/apartment-02.jpg`,
//   `img/room.jpg`
// ];
// const TITLES = [
//   `Wood and stone place`,
//   `Beautiful &amp; luxurious apartment at great location`,
//   `Canal View Prinsengracht`,
//   `Nice, cozy, warm big bed apartment`,
// ];
// const TYPES = [
//   `Apartment`,
//   `Private room`,
//   `house`,
//   `hotel`,
// ];
// const HOUSEHOLD_ITEMS = [
//   `Wi-Fi`,
//   `Washing machine`,
//   `Towels`,
//   `Heating`,
//   `Coffee machine`,
//   `Baby seat`,
//   `Kitchen`,
//   `Cabel TV`,
//   `Fridge`,
// ];
// const USER_NAMES = [
//   `Angelina`,
//   `Max`,
// ];
// const USER_AVATAR_SRC = [
//   `img/avatar-angelina.jpg`,
//   `img/avatar-max.jpg`,
// ];
// const DESCRIPTION_TEXT = [
//   `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
//   `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
// ];

// const randomBoolean = () => Boolean(Math.round(Math.random()));
// const randomNumber = ({min = 0, max = 1}) => {
//   return min + Math.floor(max * Math.random());
// };

// const getMock = () => ({
//   id: shortid.generate(),
//   city: citiesData[Math.floor(Math.random() * 5)],
//   previewSrc: IMAGES[Math.floor(Math.random() * 6)],
//   imagesSrc: IMAGES,
//   title: TITLES[Math.floor(Math.random() * 4)],
//   description: DESCRIPTION_TEXT,
//   isPremium: randomBoolean(),
//   type: TYPES[Math.floor(Math.random() * 4)],
//   raiting: randomNumber({min: 0, max: 5}),
//   bedrooms: randomNumber({min: 1, max: 3}),
//   guests: randomNumber({min: 1, max: 9}),
//   price: randomNumber({min: 0, max: 999}),
//   householdItems: new Set(HOUSEHOLD_ITEMS),
//   host: {
//     name: USER_NAMES[Math.floor(Math.random() * 2)],
//     isPro: randomBoolean(),
//     avatarUrl: USER_AVATAR_SRC[Math.floor(Math.random() * 2)],
//   },
//   location: [
//     [52.370955, 4.873096],
//     [52.379553, 4.873096],
//     [52.390955, 4.909309],
//     [52.360955, 4.909309]
//   ]
// });

// const getOffers = () => {
//   return new Array(MOCKS_TOTAL).fill(``).map(getMock);
// }

// const offers = getOffers();

const offers = [
  {
    id: shortid.generate(),
    city: {
      name: `Cologne`,
      location: [50.939485, 6.948775],
    },
    imagesSrc: [
      `img/apartment-01.jpg`,
      `img/apartment-02.jpg`,
      `img/apartment-01.jpg`,
      `img/apartment-03.jpg`,
      `img/apartment-02.jpg`,
      `img/room.jpg`
    ],
    previewSrc: `img/apartment-01.jpg`,
    title: `_`,
    price: 0,
    raiting: 0,
    type: `Apartment`,
    isPremium: true,
    bedrooms: 3,
    guests: 2,
    householdItems: new Set([
      `Wi-Fi`,
      `Washing machine`,
      `Towels`,
    ]),
    host: {
      name: `Angelina`,
      isPro: true,
      avatarUrl: `img/avatar-angelina.jpg`,
    },
    location: [50.937764, 6.961390],
  },
  {
    id: shortid.generate(),
    city: {
      name: `Paris`,
      location: [48.859335, 2.350730],
    },
    imagesSrc: [
      `img/apartment-01.jpg`,
      `img/apartment-02.jpg`,
      `img/apartment-01.jpg`,
      `img/apartment-03.jpg`,
      `img/apartment-02.jpg`,
      `img/room.jpg`
    ],
    previewSrc: `img/apartment-02.jpg`,
    title: `Wood and stone place`,
    price: 90,
    raiting: 8.35,
    type: `Private room`,
    isPremium: false,
    bedrooms: 3,
    guests: 2,
    householdItems: new Set([
      `Wi-Fi`,
      `Washing machine`,
      `Towels`,
    ]),
    host: {
      name: `Angelina`,
      isPro: true,
      avatarUrl: `img/avatar-angelina.jpg`,
    },
    location: [48.872344, 2.360418],
  },
  {
    id: shortid.generate(),
    city: {
      name: `Brussels`,
      location: [50.855889, 4.351608],
    },
    imagesSrc: [
      `img/apartment-01.jpg`,
      `img/apartment-02.jpg`,
      `img/apartment-01.jpg`,
      `img/apartment-03.jpg`,
      `img/apartment-02.jpg`,
      `img/room.jpg`
    ],
    previewSrc: `img/apartment-03.jpg`,
    title: `Canal View Prinsengracht`,
    price: 80,
    raiting: 8.2,
    type: `1`,
    isPremium: true,
    bedrooms: 3,
    guests: 2,
    householdItems: new Set([
      `Wi-Fi`,
      `Washing machine`,
      `Towels`,
    ]),
    host: {
      name: `Angelina`,
      isPro: true,
      avatarUrl: `img/avatar-angelina.jpg`,
    },
    location: [50.855889, 4.351608]
  },
  {
    id: shortid.generate(),
    city: {
      name: `Amsterdam`,
      location: [52.371512, 4.893750],
    },
    imagesSrc: [
      `img/apartment-01.jpg`,
      `img/apartment-02.jpg`,
      `img/apartment-01.jpg`,
      `img/apartment-03.jpg`,
      `img/apartment-02.jpg`,
      `img/room.jpg`
    ],
    previewSrc: `img/room.jpg`,
    title: `Nice, cozy, warm big bed apartment`,
    price: 50,
    raiting: 3,
    type: `Apartment`,
    isPremium: false,
    bedrooms: 3,
    guests: 2,
    householdItems: new Set([
      `Wi-Fi`,
      `Washing machine`,
      `Towels`,
    ]),
    host: {
      name: `Angelina`,
      isPro: true,
      avatarUrl: `img/avatar-angelina.jpg`,
    },
    location: [52.370955, 4.873096]
  },
  {
    id: shortid.generate(),
    city: {
      name: `Amsterdam`,
      location: [52.371512, 4.893750],
    },
    imagesSrc: [
      `img/apartment-01.jpg`,
      `img/apartment-02.jpg`,
      `img/apartment-01.jpg`,
      `img/apartment-03.jpg`,
      `img/apartment-02.jpg`,
      `img/room.jpg`
    ],
    previewSrc: `img/room.jpg`,
    title: `Nice, cozy, warm big bed apartment`,
    price: 50,
    raiting: 3,
    type: `Apartment`,
    isPremium: false,
    bedrooms: 3,
    guests: 2,
    householdItems: new Set([
      `Wi-Fi`,
      `Washing machine`,
      `Towels`,
    ]),
    host: {
      name: `Angelina`,
      isPro: true,
      avatarUrl: `img/avatar-angelina.jpg`,
    },
    location: [52.379553, 4.873096]
  },
];

export {offers};
