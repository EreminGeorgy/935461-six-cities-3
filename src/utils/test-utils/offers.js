const shortid = require(`shortid`);

export const offers = [
  {
    id: shortid.generate(),
    previewSrc: `img/apartment-01.jpg`,
    title: `_`,
    price: 0,
    raiting: 0,
    type: `Apartment`,
    isPremium: true,
  },
  {
    id: shortid.generate(),
    previewSrc: `img/apartment-02.jpg`,
    title: `Wood and stone place`,
    price: 90,
    raiting: 8.35,
    type: `Private room`,
    isPremium: false,
  },
  {
    id: shortid.generate(),
    previewSrc: `img/apartment-03.jpg`,
    title: `Canal View Prinsengracht`,
    price: 80,
    raiting: 8.2,
    type: `1`,
    isPremium: true,
  },
  {
    id: shortid.generate(),
    previewSrc: `img/room.jpg`,
    title: `Nice, cozy, warm big bed apartment`,
    price: 50,
    raiting: 3,
    type: `Apartment`,
    isPremium: false,
  },
];
