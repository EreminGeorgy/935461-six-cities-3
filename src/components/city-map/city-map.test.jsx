import React from "react";
import renderer from "react-test-renderer";
// import leaflet from "../../__mocks__/leaflet.js";
import {offers} from "../../utils/test-utils/offers.js";
import {CityMap} from "./city-map.jsx";

it(`<CityMap /> should render CityMap`, () => {
  const tree = renderer
    .create(<CityMap
      offers={offers}
    />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
