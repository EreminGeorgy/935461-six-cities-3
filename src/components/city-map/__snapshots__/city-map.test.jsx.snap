// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CityMap /> should render CityMap 1`] = `
<section
  className="cities__map"
>
  <div
    id="mapId"
    style={
      Object {
        "height": "580px",
      }
    }
  />
</section>
`;
